### PR TITLE
Address .NET Framework warnings for consuming projects

### LINKED ISSUE #25: Build warnings in version 1.4.0.0
If updated from 1.3.0.0 to 1.4.0.0 and see these warnings (compiling to Any CPU):

Microsoft.Common.CurrentVersion.targets(2202, 5): [MSB3246] Resolved file has a bad image, no metadata, or is otherwise inaccessible. System.BadImageFormatException: Could not load file or assembly 'nlopt_x32.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'nlopt_x32.dll' ---> System.BadImageFormatException: Could not load file or assembly 'nlopt_x32.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'nlopt_x32.dll'
...

Also for nlopt_x64. It noticed that the lib folder of the package has changed that seems to cause this. Is this an unseen side effect of a change?

Here is the quite verbose warning:

Warning MSB3246 : Resolved file has a bad image, no metadata, or is otherwise inaccessible. System.BadImageFormatException: Could not load file or assembly 'nlopt_x64.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'nlopt_x64.dll' ---> System.BadImageFormatException: Could not load file or assembly 'nlopt_x64.dll' or one of its dependencies. The module was expected to contain an assembly manifest.
File name: 'nlopt_x64.dll'
   at System.Reflection.AssemblyName.nGetFileInformation(String s)
   at System.Reflection.AssemblyName.GetAssemblyName(String assemblyFile)
   at Microsoft.Build.Shared.AssemblyNameExtension.GetAssemblyNameEx(String path)
   at Microsoft.Build.Tasks.SystemState.GetAssemblyName(String path)
   at Microsoft.Build.Tasks.ReferenceTable.SetPrimaryAssemblyReferenceItem(ITaskItem referenceAssemblyName)

### PR DESCRIPTION
**Pull request by rogerlord**
Resolves #25 

In version 1.4.0.0 a slightly different structure was chosen for the project, by multi-targeting for both .NET Framework 4.7 as well as for .NET Standard. Although the new SDK project format allows for multi-targeting, it does not seem to cater well for including native assemblies in NuGet packages. The result was that consuming projects get warnings about potential BadImageExceptions.

This PR addresses that by adding a .nuspec file which generates the NuGet package. Consumers should not have any more issues.

Version has been bumped to 1.4.1.

### CODE REVIEW

File: NLoptNet/NLoptNet.csproj [hunk 1]
```diff
@@ -57,11 +59,15 @@
     <None Include="runtimes\win-x86\native\nlopt.lib" Pack="true" PackagePath="runtimes/win-x86/native/nlopt.lib">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Include="runtimes\win-x86\native\nlopt.dll" Pack="true" PackagePath="lib/net47/nlopt_x32.dll">
```
Comment (rogerlord) [inline NLoptNet/NLoptNet.csproj:60]: These lines were causing the original warnings

File: NLoptNet/NLoptNet.csproj [hunk 2]
```diff
@@ -57,11 +59,15 @@
...
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="runtimes\win-x64\native\nlopt.dll" Pack="true" PackagePath="lib/net47/nlopt_x64.dll">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
   </ItemGroup>
-</Project>
+  
+  <ItemGroup Condition="'$(TargetFramework)'=='net47'">
```
Comment (rogerlord) [inline NLoptNet/NLoptNet.csproj:64]: When building NLoptNet, and targeting .NET 4.7, this copies the relevant nlopt DLLs to the bin folder

File: NLoptNet/build/NLoptnet.targets [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<Project>
```
Comment (rogerlord) [inline NLoptNet/build/NLoptnet.targets:1]: Runs in all consuming projects, and copies the native DLLs to the output folder when targeting .NET 4.7, 4.7.1, 4.7.2 or 4.8.

File: NLoptNet/nloptnet.nuspec [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (rogerlord) [inline NLoptNet/nloptnet.nuspec:1]: .nuspec file that works around the issues with multi-targeting experienced with the SDK csproj file

File: .github/workflows/dotnet.windows.yml [hunk 5]
```diff
@@ -17,7 +17,6 @@ jobs:
       uses: actions/setup-dotnet@v1
       with:
         dotnet-version: |
-          2.1.x
```
Comment (rogerlord) [inline .github/workflows/dotnet.windows.yml:20]: Where possible have removed the dependency on old frameworks in the Actions workflows

File: .github/workflows/dotnet.linux.yml [hunk 6]
```diff
@@ -17,13 +17,11 @@ jobs:
...
       with:
         dotnet-version: |
-          2.1.x
           3.1.x
           6.0.x
     - name: Remove netframework testing
       run: |
-        dotnet tool install --global replace-in-file 
-        replace-in-file ./NLoptNet.NuGet.Tests/NLoptNet.NuGet.Tests.csproj -set "net47;net48;" ""
+        sed -i 's/net47;net48;//g' ./NLoptNet.NuGet.Tests/NLoptNet.NuGet.Tests.csproj
```
Comment (rogerlord) [inline .github/workflows/dotnet.linux.yml:24]: Not using replace-in-file, but just directly using sed removes the need for installing .NET Core 2.1.x, which gets rid of a warning in the GitHub Action for Linux.

**Review by BrannonKing — APPROVED** [review]
I think the changes look fine overall.

**Review by BrannonKing — CHANGES_REQUESTED** [review]
The AssemblyInfo.cs file needs to have version numbers updated. I'm not sure how that relates to the GenerateAssemblyInfo flag. Also, please squash the commits. 33 is too many.

File: NLoptNet/NLoptNet.csproj [hunk 7]
```diff
@@ -1,10 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net47</TargetFrameworks>
     <RuntimeIdentifiers>linux-arm64;linux-musl-x64;linux-x64;win-x64;win-x86</RuntimeIdentifiers>
+	<TargetFrameworks>netstandard2.0;net47</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (rogerlord) [reply NLoptNet/NLoptNet.csproj:5]: Had noticed in the review of the PR, but I don't see it in my editor.

**Comment by rogerlord** [comment]
Hi @BrannonKing - I will remove the AssemblyVersion from AssemblyInfo.cs, I don't think it's needed there as it's already set elsewhere.

As for the squash - that's something you should be able to do when merging:

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

**Comment by rogerlord** [comment]
Done - removing the version in AssemblyInfo will need changing of the build, so I've just upgraded the version for now.
Tabs / spaces is also cleaned up now, in the .csproj file.

As mentioned, when you merge, if the repo is configured for that, you can squash the entire PR into one commit if you prefer to do so.

File: NLoptNet/NLoptNet.csproj (hunk 7 above)
[the remark this answers is not in the document]
Comment (rogerlord) [reply NLoptNet/NLoptNet.csproj]: Has been fixed

**Pull request merged**